### ISSUE TITLE
How to: multi-class multi-label image classification 

### ISSUE THREAD

**Issue by jozefmorvay** [deeplearning4j/deeplearning4j#4762]
As the subject suggests, I am trying to create a multi-class multi-label image classification solution. I have successfully done this in Keras, now I am in the process of migrating it to DL4j. I have run into several problems that I would to get some help with, as it would certainly speed things up for me and the documentation is frustratingly insufficient. 

I have three problems, and perhaps each one of them should be treated as a separate issue, but I want this all to be in one place for all programmers who would want to accomplish a similar/the same feat.

**Problem 1**

The images reside in directories train,valid, and test. Note that they are not further divided into categories by being placed into directories per class, as that would not make sense in a multi-label problem, although that is a possibly valid approach in single label classification. Once you open a directory test, for instance, the structure is then flat. I have a JSON file mapping an image filename onto its respective labels, i.e. : 

{"image1.jpg" : [category1, category2],
 "image2.jpg" : [category1, category3],
 "image3.jpg" : [category4, category5]} etc. 

To allow DL4j to read those class belongings, I have implemented a PathLabelGenerator:

```java
public class FilenameLabelGenerator implements PathLabelGenerator {

    private Map<String, List<String>> filenamesToLabelsMapping; //this is the JSON parsed into the map passed in as a parameter through the constructor
    private Map<String, Integer> labelsToIndexMapping; //this simply maps a class name to a number, so I can encode a multi-hot vector
    private int numberOfClasses = labelsToIndexMapping.size();

    public FilenameLabelGenerator(Map<String, List<String>> filenamesToLabelsMapping, Map<String, Integer> labelsToIndexMapping) {
        this.labelsToIndexMapping = labelsToIndexMapping;
        this.filenamesToLabelsMapping = filenamesToLabelsMapping;
    }

    @Override
    public Writable getLabelForPath(String path) {
        File file = new File(path);
        List<String> labelsForImage = filenamesToLabelsMapping.get(file.getName());
        INDArray multiHotVector = Nd4j.zeros(numberOfClasses);
        for (String label : labelsForImage) {
            multiHotVector.putScalar(labelsToIndexMapping.get(label), 1);
        }
        return new NDArrayWritable(multiHotVector);
    }

    @Override
    public Writable getLabelForPath(URI uri) {
        return getLabelForPath(new File(uri).toString());
    }
}
```
The getLabelForPath() returns NDArrayWritable as a multi-hot encoded ND4j array. In Keras, I simply create a numpy array and pass that to the model as a ground truth, but DL4j uses Writable implementations, which I do not understand the purpose of. Will this work? I have no way to test it yet, because I face other problems with the model, which I describe later. 

**Problem 2**

I read some of the Gitter archived chats, and it was mentioned multiple times that a RecordReaderMultiDataSetIterator should be used for a multi-labels classification instead of a regular RecordReaderDataSetIterator. The javadoc claims that "generate multiple inputs and multiple outputs from one or more Sequence/RecordReaders". I do not have multiple inputs, why should I use it? I generate my DataSetIterators like so:

```java
private DataSetIterator datasetIterator(List<URI> uris) {
        CollectionInputSplit inputSplit = new CollectionInputSplit(uris);
        ImageRecordReader trainRecordReader = new ImageRecordReader(HEIGHT, WIDTH, CHANNELS, labelMaker); //labelmaker is an instance of my PathLabelGenerator mentioned above
        try {
            trainRecordReader.initialize(inputSplit);
        } catch (IOException e) {
            e.printStackTrace();
        }
        //may need to use RecordReaderMultiDataSetIterator
        return new RecordReaderDataSetIterator(trainRecordReader, BATCH_SIZE, 1, NUM_CLASSES);
    }
```

**Problem 3**

For obvious reasons I use transfer learning API. So I instantiated a VGG16, and I noticed the class has its fully connected layers commented out - lines 149 through 152. Why? 

Now, I should use [multiple output layers](https://deeplearning4j.org/compgraph#multitask), so I came up with this:

```java
private ComputationGraph createVGG16TransferModel() {
        return new TransferLearning.GraphBuilder(vgg16)
                .fineTuneConfiguration(fineTuneConf)
                .setFeatureExtractor("block5_pool")
                .nOutReplace("flatten",256, WeightInit.XAVIER)
                .removeVertexAndConnections("predictions")
                .removeVertexAndConnections("fc2")
                .removeVertexAndConnections("fc1")
                .addLayer("fc1", new DenseLayer.Builder()
                        .activation(Activation.RELU)
                        .weightInit(WeightInit.XAVIER)
                        .nIn(256).nOut(256).build(), "flatten")
                .addLayer("predictions1", new OutputLayer
                        .Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .activation(Activation.SOFTMAX)
                        .nIn(256).nOut(NUM_CLASSES).build(), "fc1")
                .addLayer("predictions2", new OutputLayer
                        .Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .activation(Activation.SOFTMAX)
                        .nIn(256).nOut(NUM_CLASSES).build(), "fc1")
                .addLayer("predictions3", new OutputLayer
                        .Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .activation(Activation.SOFTMAX)
                        .nIn(256).nOut(NUM_CLASSES).build(), "fc1")
                .addLayer("predictions4", new OutputLayer
                        .Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .activation(Activation.SOFTMAX)
                        .nIn(256).nOut(NUM_CLASSES).build(), "fc1")
                .setOutputs("predictions1","predictions2","predictions3","predictions4")
                .build();
    }
```
I remove top three layers, as they are not needed, and want to train a single fully connected layer and then a separate softmax layer per class, while keep all the convolutional and pooling layers frozen. Is this a correct setup?

However, when I put a breakpoint in the code to see the VGG16 instance to inspect its vertex names to see if I have them right, Intellij is simply unable to display any information in the debugger. After expanding 'vertices' field, it just keeps displaying Collecting data... forever. Not sure what's up with that. 

EDIT: As @treo pointed out, this was caused by automatic toString() call by Intellij. After disabling it, the debugger becomes useful again.

**Comment by treo** [comment]
Concerning Problem 1:
Writables are a DataVec abstraction, i.e. they are used in your ETL pipeline, which also means that you can easily check that it works even without a running network. Simply iterate over the DataSetIterator and see if the created DataSet is what you think it should be. 

Concerning Problem 2:
You use a multi-dataset-iterator when you either need multiple inputs, multiple outputs, or both. As you have created a Network with multiple outputs you will need a multidataset iterator.

Overall it would help if you could add your keras setup, so we could tell you if your manual transformation is even close to what you actually want to do.

**Comment by AlexDBlack** [comment]
OK, so in short: On reflection, multi-class multi-label image classification isn't as easy as it should be. I'm definitely in favor of some better design/refactoring there.
You are correct in looking at PathLabelGenerator, and if you were doing *binary* multi-label it'd be perfect. But because you are doing non-binary, multi-label you need to return multiple Writable objects - 

OK, so what are your options?
1. Wait until I've refactored PathLabelGenerator to allow returning of multiple writables - and then switch to snapshots
2. Use one of the following hacks as a workaround (yes, they aren't especially elegant - but shouldn't be too hard to implement hopefully)
    a. Concatenate your labels, return a single NDArrayWritable/INDArray, then use MultiDataSetPreProcessor to split them into separate INDArrays in the MultiDataSet
    b. Extend ImageRecordReader - and override the .next() method, and just append the labels you want (no more PathLabelGenerator etc)


> it was mentioned multiple times that a RecordReaderMultiDataSetIterator should be used for a multi-labels classification instead of a regular RecordReaderDataSetIterator

Yes, for *non-binary* multi-label, RecordReaderMultiDataSetIterator is required. Why? DataSet (i.e., from RRDSI) is single in/out label only.


> I noticed the class has its fully connected layers commented out - lines 149 through 152. Why?

That's a great question...

@crockpotveggies can you explain this?
https://github.com/deeplearning4j/deeplearning4j/blob/80260f8c86a47a2821b771a23465595d310c8a46/deeplearning4j-zoo/src/main/java/org/deeplearning4j/zoo/model/VGG16.java#L147-L151



As for your architecture - looks good to me at first glance.

**Comment by AlexDBlack** [comment]
@jozefmorvay I've started working on a PR here: https://github.com/deeplearning4j/DataVec/pull/520
I'm out of time for today, but should be able to wrap that up tomorrow.

**Comment by jozefmorvay** [comment]
@treo I am including the python script here, it should be fairly easy to follow. I had to change the extension to txt lest Github wouldn't allow me to upload it.
[VGG_multiclass.txt](https://github.com/deeplearning4j/deeplearning4j/files/1806744/VGG_multiclass.txt)

@AlexDBlack Certainly, whenever you get down to it. I can help you with the PR if you wish.

As for the suggestions from both of you, I will test them later today.

**Comment by crockpotveggies** [comment]
Regarding VGG16 it's very possible that's a mistake. Our output layer
contains it's own FC layer and I may have only intended to comment out one.
I'll have a look when I'm in front of a computer.

**Comment by jozefmorvay** [comment]
@crockpotveggies I have not looked into code more thoroughly, but the VGG16 zoo model still has all the layers that it should have once instantiated, so those commented out layers are added somewhere else. It was just an off topic question, not really worth investigating IMO.

**Comment by AlexDBlack** [comment]
OK, so I've finished, tested and merged this: https://github.com/deeplearning4j/DataVec/pull/520
Snapshots for DataVec should be built soon (within hours, if not already) now that it's merged.
Snapshots info: https://deeplearning4j.org/snapshots
Updated examples for use with snapshots: https://github.com/deeplearning4j/dl4j-examples/pull/589

I'll close this issue. @jozefmorvay have a look at that PR and the javadoc + tests. I'm not 100% happy with the javadoc explaination there (with great flexibility comes many options/configurations), but hopefully you can work out how to use it. Ping me in gitter if you get stuck.
I'd like to get some multi-label image classification examples up at some point to make it easier, but no promises as to that happening soon.

I've also opened this issue to track the VGG16 question. https://github.com/deeplearning4j/deeplearning4j/issues/4800

**Comment by pawel12** [comment]
Hi,

One more question to this. Is there some simple way of evaluating multi output network?
If I try to just use ComputationGraph.evaluate() I'm getting this exception: https://github.com/deeplearning4j/deeplearning4j/blob/1bceec00a6cc7b60d873c4179da49667b5d393c0/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java#L3738
